### PR TITLE
[ie/murrtube] Fix extractor

### LINKED ISSUE #7500: [Murrtube] [MurrtubeUser] Unable to download JSON metadata: HTTP Error 404
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I'm running yt-dlp version **2023.06.22** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

JP, US, etc

### Provide a description that is worded well enough to be understood

[Murrtube] and [MurrtubeUser] extractor say they don't support URLs. No video is saved. Murrtube.net recently updated their site design so that may be the effect.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
# [Murrtube] extractor output:

yt-dlp -vU https://murrtube.net/videos/sexy-time-with-a-handsome-fox-part-1 
[debug] Command-line config: ['-vU', 'https://murrtube.net/videos/sexy-time-with-a-handsome-fox-part-1']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.06.22 [812cdfa06] (pip)
[debug] Python 3.11.4 (CPython x86_64 64bit) - macOS-10.15.7-x86_64-i386-64bit (OpenSSL 1.1.1u  30 May 2023)
[debug] exe versions: ffmpeg 6.0 (setts), ffprobe 6.0, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2023.05.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1851 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.06.22, Current version: stable@2023.06.22
yt-dlp is up to date (stable@2023.06.22)
[generic] Extracting URL: https://murrtube.net/videos/sexy-time-with-a-handsome-fox-part-1
[generic] sexy-time-with-a-handsome-fox-part-1: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] sexy-time-with-a-handsome-fox-part-1: Extracting information
[debug] Looking for embeds
ERROR: Unsupported URL: https://murrtube.net/videos/sexy-time-with-a-handsome-fox-part-1
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1555, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1631, in __extract_info
    ie_result = ie.extract(url)
                ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 708, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/generic.py", line 2568, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://murrtube.net/videos/sexy-time-with-a-handsome-fox-part-1

--------------------------------------------------
[MurrtubeUser] output:

yt-dlp --vU https://murrtube.net/gen-silver     

Usage: yt-dlp [OPTIONS] URL [URL...]

yt-dlp: error: no such option: --vU
master@MacBook-Pro ~ % yt-dlp -vU https://murrtube.net/gen-silver 
[debug] Command-line config: ['-vU', 'https://murrtube.net/gen-silver']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.06.22 [812cdfa06] (pip)
[debug] Python 3.11.4 (CPython x86_64 64bit) - macOS-10.15.7-x86_64-i386-64bit (OpenSSL 1.1.1u  30 May 2023)
[debug] exe versions: ffmpeg 6.0 (setts), ffprobe 6.0, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2023.05.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1851 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.06.22, Current version: stable@2023.06.22
yt-dlp is up to date (stable@2023.06.22)
[MurrtubeUser] Extracting URL: https://murrtube.net/gen-silver
[MurrtubeUser] gen-silver: Downloading user info
ERROR: [MurrtubeUser] gen-silver: Unable to download JSON metadata: HTTP Error 404: Not Found (caused by <HTTPError 404: 'Not Found'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 708, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/murrtube.py", line 141, in _real_extract
    data = self._download_gql(username, {
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/murrtube.py", line 42, in _download_gql
    result = self._download_json(
             ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 1062, in download_content
    res = getattr(self, download_handle.__name__)(url_or_request, video_id, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 1026, in download_handle
    res = self._download_webpage_handle(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 896, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query, expected_status=expected_status)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 853, in _request_webpage
    raise ExtractorError(errmsg, cause=err)

  File "/usr/local/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 835, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 3786, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/urllib/request.py", line 525, in open
    response = meth(req, response)
               ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/urllib/request.py", line 634, in http_response
    response = self.parent.error(
               ^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/urllib/request.py", line 563, in error
    return self._call_chain(*args)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/urllib/request.py", line 496, in _call_chain
    result = func(*args)
             ^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
```

### PR DESCRIPTION
**Pull request by DrakoCpp**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Fixes #7500 for murrtube extractor which previously was not working

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by DrakoCpp** [comment]
Hello, any news on this pull?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/murrtube.py [hunk 1]
```diff
@@ -1,164 +1,78 @@
...
+            'age_limit': 18,
             'formats': formats,
+            'description': description,
             'thumbnail': thumbnail,
-            'duration': int_or_none(meta.get('duration')),
-            'uploader': try_get(meta, lambda x: x['user']['name']),
-            'view_count': meta.get('viewsCount'),
-            'like_count': meta.get('likesCount'),
-            'comment_count': meta.get('commentsCount'),
-            'tags': meta.get('tagList'),
```
[the remark this answers is not in the document]
Comment (DrakoCpp) [reply yt_dlp/extractor/murrtube.py:98]: Added view, like and comment. Tags are not available anymore.

File: yt_dlp/extractor/murrtube.py [hunk 2]
```diff
@@ -1,164 +1,78 @@
...
-            'age_limit': 18,
+            'uploader': uploader,
         }
-
-
-class MurrtubeUserIE(MurrtubeIE):  # XXX: Do not subclass from concrete IE
```
[the remark this answers is not in the document]
Comment (DrakoCpp) [reply yt_dlp/extractor/murrtube.py:103]: Added back

**Comment by DrakoCpp** [comment]
@pukkandan it should be ready

**Review by bashonly — CHANGES_REQUESTED** [review]
I've made several review comments and Github has auto-hidden some of them, please make sure to read & address all of them

File: yt_dlp/extractor/murrtube.py [hunk 3]
```diff
@@ -1,116 +1,104 @@
...
-    }
+    _TESTS = [
+        {
+            'url': 'https://murrtube.net/stormy',
+            'info_dict': {
+                'id': 'stormy',
+            },
+            'playlist_mincount': 27,
+        }
+    ]
```
[the remark this answers is not in the document]
Comment (DrakoCpp) [reply yt_dlp/extractor/murrtube.py:105]: If i revert this to `_TEST` i get the following error:

```
AssertionError: MurrtubeUserIE has _TEST and _TESTS
```

File: yt_dlp/extractor/murrtube.py [hunk 4]
```diff
@@ -1,116 +1,104 @@
...
-        data = self._download_gql(video_id, {
-            'operationName': 'Medium',
-            'variables': {
-                'id': video_id,
+    _TESTS = [
+        {
+            'url': 'https://murrtube.net/videos/inferno-x-skyler-148b6f2a-fdcc-4902-affe-9c0f41aaaca0',
+            'md5': '70380878a77e8565d4aea7f68b8bbb35',
+            'info_dict': {
+                'id': 'ca885d8456b95de529b6723b158032e11115d',
```
[the remark this answers is not in the document]
Comment (DrakoCpp) [reply yt_dlp/extractor/murrtube.py:32]: No, they seem to not be available anymore after the site revamp

File: yt_dlp/extractor/murrtube.py [hunk 5]
```diff
@@ -1,116 +1,104 @@
...
+    def _real_initialize(self):
+        homepage = self._download_webpage(
+            'https://murrtube.net', None, note='Getting session token')
+        data = self._hidden_inputs(homepage)
+        self._download_webpage(
+            'https://murrtube.net/accept_age_check', None, 'Set age cookie', data=urlencode_postdata(data))
 
+    def _real_extract(self, url):
+        video_id = self._match_valid_url(url)
+        video_page = self._download_webpage(url, video_id)
```
[the remark this answers is not in the document]
Comment (DrakoCpp) [reply yt_dlp/extractor/murrtube.py:75]: So it's ok this way?

File: yt_dlp/extractor/murrtube.py [hunk 6]
```diff
@@ -1,116 +1,109 @@
...
-
-        if determine_ext(format_url) == 'm3u8':
-            formats = self._extract_m3u8_formats(
-                format_url, video_id, 'mp4', entry_protocol='m3u8_native', fatal=False)
-        else:
-            formats = [{'url': format_url}]
+        video_page = self._download_webpage(url, video_id)
+        video_attrs = extract_attributes(get_element_html_by_id('video', video_page))
+        playlist = update_url(video_attrs['data-url'], query=None)
+        video_id = self._search_regex(r'/([^/?#]+)/index.m3u8', playlist, 'video id')
```
Comment (bashonly) [inline yt_dlp/extractor/murrtube.py:78]:
```suggestion
        video_id = self._search_regex(r'/([\da-f]+)/index.m3u8', playlist, 'video id')
```

**Review by bashonly — APPROVED** [review]

**Pull request merged**